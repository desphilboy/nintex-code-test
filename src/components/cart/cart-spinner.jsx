import React from 'react';


export const CartSpinner = () => <div> Loading ...</div>;

export default CartSpinner;