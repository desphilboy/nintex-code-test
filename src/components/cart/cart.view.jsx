import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';


const TwoColumnContainer = styled.div`
	display: flex;
	flex-direction: row;
`;

const Section = styled.section`
	margin: 20px auto;
	padding: 4px;
	background: MintCream;
	font-size: 1em;
	border: 4px solid LightCyan;
	border-radius: 20px;
	display: flex;
	flex-direction: column;
	justifyContent: 'space-between';
	min-width: 200px;
	max-width: 300px;
	margin: 20px;
	padding 10px;

	form {
		display: flex;
  		label {
  			margin: 4px;
  			min-width: 100px;
  		}

	input {
		margin: 4px;
		display: flex;
		flex-grow: 1;
		min-width: 100px;
	}
	button {
		margin: 4px;
		min-width: 80px;
	}

}`;

const ProductPill = styled.div`
	background: FloralWhite;
	display: flex;
	flex-direction: column;
	border: 1px solid LightGray;
	border-radius: 15px;
	min-height: 100px;
	width: 80%;
	margin: 10px auto;
	padding: 15px;
	p {
		margin: auto;
		display: block;
	}
`;

export const CartView = ({ products, addProduct, orders, removeProduct, discounts }) => {
	const [orderList, setOrderList] = useState({});
	useEffect(() => setOrderList(orders));

	return(
		<TwoColumnContainer>
			<Section>
			<h3> Products:</h3>
				{products && products.data && products.data.map((product, index) => (
					<ProductPill key={`product-${index}`}>
						<p>{product.name}</p>
						<p>{product.formattedPrice}</p>
						<button className="addButton" onClick={addProduct(product)}>Add to cart</button>
					</ProductPill>
					))}
			</Section>
			<Section>
			<h3> Cart: </h3>
				{Object.keys(orders || {}).map((key) => ( orders[key].num > 0 &&
					<ProductPill key={`order-${key}`}>
						<p>{orders[key].name}</p>
						<p>{orders[key].formattedPrice}</p>
						<p>No: {orders[key].num}</p>
						<p>total: {orders[key].num * orders[key].price}</p>
						<button className="removeButton" onClick={removeProduct(orders[key])}>Remove from cart</button>
					</ProductPill>
					))}
				<h3>Total before discount: {Object.keys(orders).reduce((total, key) => total + orders[key].num * orders[key].price, 0)}</h3>
				<h3>Discount: {discounts(Object.keys(orders).reduce((total, key) => total + orders[key].num * orders[key].price, 0))}</h3>
				<h3>Payable amount:{
					Object.keys(orders).reduce((total, key) => total + orders[key].num * orders[key].price, 0) - discounts(Object.keys(orders).reduce((total, key) => total + orders[key].num * orders[key].price, 0))
				}</h3>
			</Section>
		</TwoColumnContainer>
	);
};


CartView.propTypes = {
	discounts: PropTypes.func,
	orders: PropTypes.object,
	addProduct: PropTypes.func,
	removeProduct: PropTypes.func,
	products: PropTypes.object,
};

CartView.defaultProps = {
	discounts: () => 0,
	orders: {},
	addProduct: () => {},
	removeProduct: () => {},
	products: {},
};

export default CartView;