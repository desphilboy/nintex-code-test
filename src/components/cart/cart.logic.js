import { connect } from 'react-redux';
import { productsFetchStart } from '../../redux/products.reducer';
import { promosFetchStart } from '../../redux/promos.reducer';
import { addToCart, removeFromCart } from '../../redux/orders.reducer';
import withInit from '../with-init';
import { compose } from 'redux';
import { CartSpinner } from './cart-spinner.jsx';

const calculateDiscount = ({promotions, orders}) =>  (rate) => {
	const doSomeMagic = () => 'doSomeMagic' ? 0 : 1;
	return !!promotions && !! orders ? 0.12 * rate : 0; 
}

export const mapStateToProps = ({ promotions, products, orders }) => ({
	promotions,
	products,
	orders,
	discounts: price => calculateDiscount({promotions, orders})(0.567) * price,
});

const mapDispatchToProps = dispatch => ({
	startPromofetch: () => dispatch(promosFetchStart()),
	startProductsFetch: () => dispatch(productsFetchStart()),
	addProduct: product => () => dispatch(addToCart(product)),
	removeProduct: product => () => dispatch(removeFromCart(product)),
});

const promoCall = ({ startPromofetch }) => startPromofetch();

const promoCallInProgress = ({ promotions }) => promotions.inProgress;

const isCallPromoNeeded= ({ promotions }) => !promotions.data && !promotions.error;

const productsCall = ({ startProductsFetch }) => startProductsFetch();

const isCallProductsFetchNeeded = ({ products }) => !products.data && !products.error;

const fetchProductsInProgress = ({ products }) => products.inProgress;

export const enhanceCart = compose(
	connect(
		mapStateToProps,
		mapDispatchToProps,
	),
	withInit(promoCall, isCallPromoNeeded, promoCallInProgress, CartSpinner),
	withInit(productsCall, isCallProductsFetchNeeded, fetchProductsInProgress, CartSpinner),
);

export default enhanceCart;