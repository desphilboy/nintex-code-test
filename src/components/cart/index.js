import { CartView } from './cart.view.jsx';
import { enhanceCart } from './cart.logic';


export default enhanceCart(CartView);
