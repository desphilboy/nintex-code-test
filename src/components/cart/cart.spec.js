
import React from 'react';
import { shallow } from 'enzyme';
import { CartView } from './cart.view';
import Spinner from './cart-spinner';
import * as logic from './cart.logic';

describe('component Cart', () => {

	describe('CartView', () => {
		beforeEach(() => {
			jest.clearAllMocks();
		});

		it('matches the snapshot with default props', () => {
			const wrapper = shallow(<CartView/>);

			expect(wrapper).toMatchSnapshot();
		});

		it('matches the snapshot with data', () => {
			const wrapper = shallow(<CartView
				products={{ data: [
					{ id: 'wf', name: 'Workflow', price: 199.99, formattedPrice: '$199.99' },
					{ id: 'docgen', name: 'Document Generation', price: 9.99, formattedPrice: '$9.99' },
					{ id: 'form', name: 'Form', price: 99.99, formattedPrice: '$99.99'},
				]}}

			/>);

			expect(wrapper).toMatchSnapshot();
		});

		it('calls addProduct if add button is clicked', () => {
			const addProductSpy = jest.fn();
			const wrapper = shallow(<CartView
				products={{ data: [
					{ id: 'wf', name: 'Workflow', price: 199.99, formattedPrice: '$199.99' },
				]}}
				addProduct={addProductSpy}
			/>);
			wrapper.find('button.addButton').simulate('click');
			expect(addProductSpy).toHaveBeenCalled();
		});
	});

	describe('Spinner', () => {
		it('matches the snapshot', () => {
			const wrapper = shallow(<Spinner/>);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('mapStateToProps', () => {
		it('brings promotions and products from state', () => {
			const testState = {
				promotions: 'aloo',
				products: 'baloo',
				orders: 'maloo',
			};

			const map = logic.mapStateToProps(testState);

			expect(map.promotions).toBe('aloo');
			expect(map.products).toBe('baloo');
			expect(map.orders).toBe('maloo');
		});
	});
});