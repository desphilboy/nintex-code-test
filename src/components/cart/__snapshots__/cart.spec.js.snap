// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Cart CartView matches the snapshot with data 1`] = `ShallowWrapper {}`;

exports[`component Cart CartView matches the snapshot with default props 1`] = `ShallowWrapper {}`;

exports[`component Cart Spinner matches the snapshot 1`] = `ShallowWrapper {}`;
